custom-recipes/named-entity-recognition-extract/recipe.json

```diff
@@ -152,6 +153,26 @@
                 }
             ],
             "defaultValue": "standard"
+        },
+        {
+            "visibilityCondition": "model.ner_model=='flair' && model.advanced_settings",
+            "name": "flair_device",
+            "label": "Processing",
+            "type": "SELECT",
+            "wt1Loggable": true,
+            "description": "CPU/GPU",
+            "selectChoices": [
+                {
+                    "value": "cpu",
+                    "label": "CPU"
+                },
+                {
+                    "value": "cuda",
+                    "label": "CUDA"
+                }
+
+            ],
+            "defaultValue": "cpu"
```

Comment:
@avigny
*I don't dynamically add the cuda option for two reasons:
  * First and foremost, it's quite hard to know what will be the runtime env when the dynamic select is computed (e.g. k8s). I initially thought about using `torch.cuda.is_available` but that would not work.
  * I guess it's good that the user sees the option is available (ideally it could have been greyed out but not sure that's feasible with plugins "autoparams")

* I simply use `cuda` and let torch chooses the default device in case there are multiple ones. I think it's enough here (not even sure people will use this option tbh, at least we will see that with wt1Loggable) but let me know if you think it's worth giving more flexible options to the user.

WDYT?

Reply:
I agree, I don't see an easy way to know if a GPU will be available. you could  (or not) have one available from your design node and have a completely different situation in the env running the recipe (k8s for example as you say).

If the user specifies `CPU` he will have the recipe run on CPU as expected
If the user specifies `CUDA` he expects the recipe to run on GPU (if possible)

I'd say that selecting CUDA means that we will try our best to use a GPU if available. if no GPU is detected or found we should fallback to use CPU (with a warning in the logs) and not fail. (I expect and hope that flair does this)

Reply:
> I'd say that selecting CUDA means that we will try our best to use a GPU if available. if no GPU is detected or found we should fallback to use CPU (with a warning in the logs) and not fail. (I expect and hope that flair does this)

Probably I can wrap the `flair.device = xxx` with a proper `if torch.cuda.is_available` at runtime

Thanks for your input :+1:

Reply:
currently it looks like this will fail:
```python
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/jobs/CHATGPTTESTING/Build_NER_summary_2023-06-28T13-00-15.319/compute_NER_summary_NP/custom-python-recipe/pyout5WustIIDjJbN/python-exec-wrapper.py", line 204, in <module>
[15:00:24] [INFO] [dku.utils]  -     exec(f.read())
[15:00:24] [INFO] [dku.utils]  -   File "<string>", line 77, in <module>
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/plugins/dev/named-entity-recognition/python-lib/dku_io_utils.py", line 80, in process_dataset_chunks
[15:00:24] [INFO] [dku.utils]  -     output_df = func(df=df, **kwargs)
[15:00:24] [INFO] [dku.utils]  -   File "<string>", line 64, in compute_entities_df
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/plugins/dev/named-entity-recognition/python-lib/ner/flair.py", line 45, in extract_entities
[15:00:24] [INFO] [dku.utils]  -     tagger = get_model(model_id)
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/plugins/dev/named-entity-recognition/python-lib/ner/flair.py", line 41, in get_model
[15:00:24] [INFO] [dku.utils]  -     return SequenceTagger.load(model_path)
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/code-envs/python/plugin_named-entity-recognition_managed/lib/python3.7/site-packages/flair/nn/model.py", line 144, in load
[15:00:24] [INFO] [dku.utils]  -     model = cls._init_model_with_state_dict(state)
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/code-envs/python/plugin_named-entity-recognition_managed/lib/python3.7/site-packages/flair/models/sequence_tagger_model.py", line 640, in _init_model_with_state_dict
[15:00:24] [INFO] [dku.utils]  -     **kwargs,
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/code-envs/python/plugin_named-entity-recognition_managed/lib/python3.7/site-packages/flair/nn/model.py", line 74, in _init_model_with_state_dict
[15:00:24] [INFO] [dku.utils]  -     model = cls(**kwargs)
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/code-envs/python/plugin_named-entity-recognition_managed/lib/python3.7/site-packages/flair/models/sequence_tagger_model.py", line 194, in __init__
[15:00:24] [INFO] [dku.utils]  -     self.crf = CRF(self.label_dictionary, self.tagset_size, init_from_state_dict)
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/code-envs/python/plugin_named-entity-recognition_managed/lib/python3.7/site-packages/flair/models/sequence_tagger_utils/crf.py", line 33, in __init__
[15:00:24] [INFO] [dku.utils]  -     self.to(flair.device)
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/code-envs/python/plugin_named-entity-recognition_managed/lib/python3.7/site-packages/torch/nn/modules/module.py", line 852, in to
[15:00:24] [INFO] [dku.utils]  -     return self._apply(convert)
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/code-envs/python/plugin_named-entity-recognition_managed/lib/python3.7/site-packages/torch/nn/modules/module.py", line 552, in _apply
[15:00:24] [INFO] [dku.utils]  -     param_applied = fn(param)
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/code-envs/python/plugin_named-entity-recognition_managed/lib/python3.7/site-packages/torch/nn/modules/module.py", line 850, in convert
[15:00:24] [INFO] [dku.utils]  -     return t.to(device, dtype if t.is_floating_point() or t.is_complex() else None, non_blocking)
[15:00:24] [INFO] [dku.utils]  -   File "/Users/avigny/Library/DataScienceStudio/dss_home/code-envs/python/plugin_named-entity-recognition_managed/lib/python3.7/site-packages/torch/cuda/__init__.py", line 166, in _lazy_init
[15:00:24] [INFO] [dku.utils]  -     raise AssertionError("Torch not compiled with CUDA enabled")
[15:00:24] [INFO] [dku.utils]  - AssertionError: Torch not compiled with CUDA enabled
```

Reply:
Looking at your logs, you are on macos and that's expected. I will display the option only for supported platform 👍

Reply:
here is a snippet that could be useful:
https://github.com/dataiku/dip/blob/f0ffd2c373957ac4a495322fbd7155d4b4a0936e/src/main/python/dataiku/doctor/prediction/multi_layer_perceptron_models.py#L109-L114

Reply:
> Looking at your logs, you are on macos and that's expected. I will display the option only for supported platform 👍

Actually applies as for runtime env so I have to display it anyway ^^. Hopefully the wrapping will avoid the issue you got

Reply:
for UX consistence, I asked @tmwly (gpu-experience) and he advised to rather let it fail (as a user I quite understand that you'd rather have explicit failures => that would avoid questions to the support like the one you had recently), do you see any objections?

Reply:
No objection!
this is not hard to change in the future if necessary, so we can stick with the current choice!
👍